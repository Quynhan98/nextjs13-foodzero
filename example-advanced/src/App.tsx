// import { CodeSplitting } from './components/code-splitting';
import { ComponentParent } from './components/context/component-context';
import { CountProvider } from './components/context/count-context';
import { ExampleCustomHook } from './components/custom-hook';
import { ExampleErrorBoundary } from './components/error-boundary';
import { ForwardingRefExample } from './components/forwarding-ref';
import { ExampleFragment } from './components/fragment';
import { OptimizingPerformance } from './components/optimizing-performance';
import { ExampleProfiler } from './components/profiler';
import { ExampleSWR } from './components/swr';
import { UncontrolledComponent } from './components/uncontroled-component';

const App = (): JSX.Element => {
  return (
    <div className="App">
      {/* <CodeSplitting /> */}
      <ComponentParent />
      <CountProvider />
      <ExampleErrorBoundary />
      <UncontrolledComponent />
      <ExampleProfiler />
      <ForwardingRefExample />
      <OptimizingPerformance />
      <ExampleCustomHook />
      <ExampleFragment />
      <ExampleSWR />
    </div>
  );
};

export default App;
