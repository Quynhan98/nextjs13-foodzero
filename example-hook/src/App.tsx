import { ExampleUseCallback } from './components/useCallback';
import { ExampleUseContext } from './components/useContext';
import { ExampleUseEffect } from './components/useEffect';
import { ExampleUseLayoutEffect } from './components/useLayoutEffect';
import { ExampleUseMemo } from './components/useMemo';
import { ExampleUseReducer } from './components/useReducer';
import { ExampleUseRef } from './components/useRef';
import { ExampleUseState } from './components/useState';

function App() {
  return (
    <div className="App">
      <ExampleUseLayoutEffect />
      <ExampleUseState />
      <ExampleUseEffect />
      <ExampleUseContext />
      <ExampleUseRef />
      <ExampleUseReducer />
      <ExampleUseCallback />
      <ExampleUseMemo />
    </div>
  );
}

export default App;
