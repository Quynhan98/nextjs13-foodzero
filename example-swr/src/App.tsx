import './styles/reset.css';
import { Route, Routes } from 'react-router-dom';
import { Header } from './components/header';
import { ListMovie } from './pages/movies-page';
import SingleMoviePage from './pages/single-page';

const App = (): JSX.Element => {
  return (
    <>
      <Header />
      <Routes>
        <Route path="/" element={<ListMovie />}>
          <Route path=":movieType" element={<ListMovie />}></Route>
        </Route>
        <Route path="movie/:id" element={<SingleMoviePage />}></Route>
      </Routes>
    </>
  );
};

export default App;
