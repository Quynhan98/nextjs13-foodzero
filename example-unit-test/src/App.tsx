import { Fetch } from './example/fetch';
import { HiddenMessage } from './example/hidden-message';
import { User } from './example/user';

const App = (): JSX.Element => {
  return (
    <div className="App">
      <HiddenMessage>Example Unit-Testing</HiddenMessage>
      <Fetch />
      <User />
    </div>
  );
};

export default App;
