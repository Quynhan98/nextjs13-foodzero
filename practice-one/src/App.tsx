import '../node_modules/bootstrap/dist/css/bootstrap.min.css';
import './styles/main.css';
import { Header } from '@components/header';
import { MainService } from '@components/main-service';
import { Footer } from '@components/footer';

function App() {
  return (
    <div className="App">
      <Header />
      <MainService />
      <Footer />
    </div>
  );
}

export default App;
