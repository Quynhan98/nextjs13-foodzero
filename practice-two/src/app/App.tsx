import '../styles/main.css';
import { ChangeEvent, useCallback, useState } from 'react';
import { SERVER_MESSAGES } from '@constant/messages';
import URL_PAGE from '@constant/url';
import { formValidate, ValidationResult } from '@helpers/validation-form';
import { deleteData, postData, updateData } from '@services/fetch-api';
import useGetData from '@hooks/use-get-data';
import { useDebounce } from '@hooks/use-debounce';
import { Book } from '@interface/book';
import { Error } from '@components/common/error';
import { Popup } from '@components/common/popup';
import Search from '@components/common/search';
import { Form } from '@components/form';
import { Header } from '@components/header';
import { Books } from '@components/list-book';
import { SearchEmptyResult } from '@components/common/search-empty';

export function TestComponent() {
  const [value, setValue] = useState(0);

  const inc = () => {
    setValue(value + 1);
  };

  const debouncedValue = useDebounce(value, 1000);
  return (
    <div>
      <button onClick={inc}>Increment</button>
      <span data-testid="debouncedValue">{debouncedValue}</span>
      <span data-testid="value">{value}</span>
    </div>
  );
}

const App = (): JSX.Element => {
  const [book, setBook] = useState<Book | undefined>();
  const [isOpen, setIsOpen] = useState(false);
  const [search, setSearch] = useState('');
  const searchValue = useDebounce(search, 500);
  const { data, error, setData, setError } = useGetData<Book>(`${URL_PAGE}${searchValue}`);

  const handelToggleForm = (): void => {
    setIsOpen((prevIsOpen) => (prevIsOpen = !prevIsOpen));

    if (isOpen) {
      setBook(undefined);
      document.body.style.overflow = '';
    } else {
      document.body.style.overflow = 'hidden';
    }
  };

  const handleCreate = async (book: Book): Promise<void> => {
    try {
      const validateFormCreate: ValidationResult = formValidate(book);

      if (validateFormCreate.isValid) {
        await postData(book, URL_PAGE);

        setData([book, ...data]);
        handelToggleForm();
      }
    } catch (error) {
      setError(SERVER_MESSAGES.SERVER_RESPONSE_ERROR);
      handelToggleForm();
    }
  };

  const handleDelete = useCallback(
    async (id: number): Promise<void> => {
      try {
        await deleteData(id, URL_PAGE);
        const bookArr: Book[] = data.filter((item) => item.id !== id);

        setData(bookArr);
      } catch (error) {
        setError(SERVER_MESSAGES.SERVER_DELETE_ERROR);
      }
    },
    [data, setData, setError]
  );

  const clickButtonEdit = (book: Book): void => {
    setBook(book);
    handelToggleForm();
  };

  const handleEditBook = async (book: Book): Promise<void> => {
    try {
      const validateFormEdit: ValidationResult = formValidate(book);

      if (validateFormEdit.isValid) {
        await updateData(book.id, URL_PAGE, book);

        const newBooks: Book[] = data.map((oldBook) => {
          if (oldBook.id === book.id) {
            return book;
          } else {
            return oldBook;
          }
        });

        setData(newBooks);
        handelToggleForm();
      }
    } catch (error) {
      setError(SERVER_MESSAGES.SERVER_EDIT_ERROR);
    }
  };

  const closeError = (): void => {
    setError('');
  };

  const handleChangeSearch = (e: ChangeEvent<HTMLInputElement>) => {
    const searchTerm: string = e.target.value;

    setSearch(`?title_like=${searchTerm}`);
  };

  return (
    <div className="container" data-testid="app">
      <Header toggleForm={handelToggleForm} />
      <Search onChange={handleChangeSearch} />
      {searchValue && data.length === 0 ? (
        <SearchEmptyResult />
      ) : (
        <Books onClickButtonEdit={clickButtonEdit} handleRemove={handleDelete} books={data} />
      )}
      {isOpen && (
        <Popup handleClose={handelToggleForm}>
          <Form
            onHandleEdit={handleEditBook}
            onCreate={handleCreate}
            onHandleClose={handelToggleForm}
            selectedBook={book}
          />
        </Popup>
      )}
      {error && <Error onHandleClick={closeError}>{error}</Error>}
      <TestComponent />
    </div>
  );
};

export default App;
