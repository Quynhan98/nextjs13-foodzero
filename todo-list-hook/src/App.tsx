import './styles/reset.css';
import './styles/main.css';
import Todos from './component/main-service/todos';

const App = (): JSX.Element => {
  return (
    <div className="todoapp">
      <Todos />
    </div>
  );
};
export default App;
