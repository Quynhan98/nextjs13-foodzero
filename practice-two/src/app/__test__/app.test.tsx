import { fireEvent, render, screen } from '@testing-library/react';
import App from '@app/App';
import useGetData from '@hooks/use-get-data';
import * as React from 'react';

jest.mock('@hooks/use-get-data', () => jest.fn());

describe('Test App', () => {
  beforeEach(() => {
    (useGetData as jest.Mock).mockImplementation(() => ({
      data: [
        {
          id: 1,
          title: 'title',
          author: 'nhan',
          price: 123,
          desc: 'this is a book',
          image: 'url',
        },
      ],
      error: 'Error Message',
    }));
  });

  // TODO
  // - Implement unit test for closeError function
  // NEED
  // - I need a solution to this problem
  const setErrorMock = jest.fn();
  const useStateSpy: any = jest.spyOn(React, 'useState');
  let button: HTMLButtonElement;

  useStateSpy.mockImplementation((init: any) => [init, setErrorMock]);

  test('App call setError', async () => {
    const { getByTestId } = render(<App />);

    button = getByTestId('btn-close-err') as HTMLButtonElement;
    fireEvent.click(button);
    expect(setErrorMock).toHaveBeenCalled();
  });

  test('App matchers DOM Snapshot', () => {
    const component = render(<App />);
    expect(component).toMatchSnapshot();
  });

  test('App should call Form Create', () => {
    render(<App />);

    fireEvent.click(screen.getByText('Add New Book'));
    expect(screen.getByText('Create New Book')).toHaveTextContent('Create New Book');
    fireEvent.click(screen.getByText('Cancel'));
    expect(screen.getByTestId('app')).not.toHaveTextContent('Create New Book');
  });

  test('App should call Search', () => {
    render(<App />);

    fireEvent.change(screen.getByTestId('search'), { target: { value: 'Book' } });
    expect(screen.getByTestId('search')).toHaveValue('Book');
  });

  test('App Click Edit toggle form', () => {
    render(<App />);

    fireEvent.click(screen.getAllByText('Edit')[0]);
    expect(screen.getByTestId('app')).toHaveTextContent('Edit Book');
  });

  test('App should call handle create', () => {
    render(<App />);

    fireEvent.click(screen.getByText('Add New Book'));
    expect(screen.getByText('Create New Book')).toHaveTextContent('Create New Book');
    fireEvent.click(screen.getByText('Create'));
  });

  //
  test('App handle edit', async () => {
    render(<App />);
    fireEvent.click(screen.getAllByRole('button', { name: /Edit/i })[0]);
  });
});
