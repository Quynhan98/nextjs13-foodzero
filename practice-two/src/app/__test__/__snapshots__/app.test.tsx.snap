// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test App App matchers DOM Snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
        data-testid="app"
      >
        <header
          class="page-header"
        >
          <h2
            class="heading"
          >
            A list of books
          </h2>
          <div
            class="nav-btn"
          >
            <button
              class="button btn-primary btn-large"
              type="button"
            >
              Add New Book
            </button>
          </div>
        </header>
        <div
          class="filter-search"
        >
          <input
            class="search-input"
            data-testid="search"
            placeholder="Search..."
            type="search"
          />
        </div>
        <div
          class="book-list"
        >
          <div
            class="card-group"
          >
            <div
              class="card"
            >
              <div
                class="card-body"
              >
                <img
                  alt="image title"
                  class="card-image"
                  src="url"
                />
                <h3
                  class="card-title"
                >
                  title
                </h3>
                <span
                  class="card-author"
                >
                  nhan
                </span>
                <span
                  class="card-price"
                  data-testid="price"
                >
                  Price: 
                  123
                   USD
                </span>
                <p
                  class="card-desc"
                >
                  this is a book
                </p>
                <div
                  class="button-group"
                >
                  <button
                    class="button btn-secondary btn-small"
                    type="button"
                  >
                    Edit
                  </button>
                  <button
                    class="button btn-warning btn-small"
                    type="button"
                  >
                    Delete
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="notification-error"
          data-testid="error"
        >
          <button
            class="icon-close__error"
            data-testid="btn-close-err"
          >
            x
          </button>
          Error Message
        </div>
        <div>
          <button>
            Increment
          </button>
          <span
            data-testid="debouncedValue"
          >
            0
          </span>
          <span
            data-testid="value"
          >
            0
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
      data-testid="app"
    >
      <header
        class="page-header"
      >
        <h2
          class="heading"
        >
          A list of books
        </h2>
        <div
          class="nav-btn"
        >
          <button
            class="button btn-primary btn-large"
            type="button"
          >
            Add New Book
          </button>
        </div>
      </header>
      <div
        class="filter-search"
      >
        <input
          class="search-input"
          data-testid="search"
          placeholder="Search..."
          type="search"
        />
      </div>
      <div
        class="book-list"
      >
        <div
          class="card-group"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <img
                alt="image title"
                class="card-image"
                src="url"
              />
              <h3
                class="card-title"
              >
                title
              </h3>
              <span
                class="card-author"
              >
                nhan
              </span>
              <span
                class="card-price"
                data-testid="price"
              >
                Price: 
                123
                 USD
              </span>
              <p
                class="card-desc"
              >
                this is a book
              </p>
              <div
                class="button-group"
              >
                <button
                  class="button btn-secondary btn-small"
                  type="button"
                >
                  Edit
                </button>
                <button
                  class="button btn-warning btn-small"
                  type="button"
                >
                  Delete
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="notification-error"
        data-testid="error"
      >
        <button
          class="icon-close__error"
          data-testid="btn-close-err"
        >
          x
        </button>
        Error Message
      </div>
      <div>
        <button>
          Increment
        </button>
        <span
          data-testid="debouncedValue"
        >
          0
        </span>
        <span
          data-testid="value"
        >
          0
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
